go/vt/vtgate/planbuilder/operators/horizon_planning.go

```diff
@@ -19,76 +19,179 @@ package operators
+	p *Projection,
+	src *Vindex,
+) (ops.Operator, rewrite.ApplyResult, error) {
+	for _, column := range p.Columns {
+		expr := column.GetExpr()
+		_, _, err := src.AddColumn(ctx, aeWrap(expr))
+		if err != nil {
+			return nil, false, err
+		}
 	}
+	return src, rewrite.NewTree, nil
+}
+
+// pushDownProjectionInApplyJoin pushes down a projection operation into an ApplyJoin operation.
+// It processes each input column and creates new JoinColumns for the ApplyJoin operation based on
+// the input column's expression. It also creates new Projection operators for the left and right
+// children of the ApplyJoin operation, if needed.
+func pushDownProjectionInApplyJoin(ctx *plancontext.PlanningContext, p *Projection, src *ApplyJoin) (ops.Operator, rewrite.ApplyResult, error) {
+	var (
+		// idx is the current column index in the Projection's columns.
+		// in is the current column's ProjExpr.
+		idx int
+		in  ProjExpr
 
-	sel, isSel := in.Select.(*sqlparser.Select)
+		// lhsCols and rhsColumns store the columns to be added to the left and right child projections.
+		// lhsNames and rhsNames store the corresponding column names.
+		lhsCols, rhsColumns []ProjExpr
+		lhsNames, rhsNames  []string
+	)
+
+	// Iterate through each column in the Projection's columns.
+	for idx = 0; idx < len(p.Columns); idx++ {
+		in = p.Columns[idx]
+		expr := in.GetExpr()
+
+		// Check if the current expression can reuse an existing column in the ApplyJoin.
+		if _, found := canReuseColumn(ctx, src.ColumnsAST, expr, jcToExpr); found {
+			continue
+		}
+
+		// Get a JoinColumn for the current expression.
+		col, err := src.getJoinColumnFor(ctx, expr)
+		if err != nil {
+			return nil, false, err
+		}
+
+		// Update the left and right child columns and names based on the JoinColumn type.
+		switch {
+		case col.IsPureLeft():
+			lhsCols = append(lhsCols, in)
+			lhsNames = append(lhsNames, p.ColumnNames[idx])
+		case col.IsPureRight():
+			rhsColumns = append(rhsColumns, in)
+			rhsNames = append(rhsNames, p.ColumnNames[idx])
+		case col.IsMixedLeftAndRight():
+			for _, lhsExpr := range col.LHSExprs {
+				lhsCols = append(lhsCols, &Expr{E: lhsExpr})
+				lhsNames = append(lhsNames, sqlparser.String(lhsExpr))
+			}
+
+			rhsColumns = append(rhsColumns, &Expr{E: col.RHSExpr})
+			rhsNames = append(rhsNames, p.ColumnNames[idx])
+		}
+
+		// Add the new JoinColumn to the ApplyJoin's ColumnsAST.
+		src.ColumnsAST = append(src.ColumnsAST, col)
+	}
+
+	// Create and update the Projection operators for the left and right children, if needed.
+	if len(lhsCols) > 0 {
+		lhsProj, err := createProjection(src.LHS)
+		if err != nil {
+			return nil, false, err
+		}
+		lhsProj.ColumnNames = lhsNames
+		lhsProj.Columns = lhsCols
+		src.LHS = lhsProj
+	}
+	if len(rhsColumns) > 0 {
+		rhsProj, err := createProjection(src.RHS)
+		if err != nil {
+			return nil, false, err
+		}
+		rhsProj.ColumnNames = rhsNames
+		rhsProj.Columns = rhsColumns
+		src.RHS = rhsProj
```

Comment:
It would be great to extract this into a single method that can be reused for both `lhsCols` and `rhsColumns`, i think it would be easier to maintain :) no strong opinion tho